controllers/watcher_controller.go

```diff
@@ -69,35 +71,45 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 	watcherObj := &v1alpha1.Watcher{}
 	if err := r.Get(ctx, client.ObjectKey{Name: req.Name, Namespace: req.Namespace}, watcherObj); err != nil {
 		logger.Error(err, "Failed to get reconciliation object")
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		return ctrl.Result{Requeue: true}, client.IgnoreNotFound(err)
 	}
 
 	// check if deletionTimestamp is set, retry until it gets fully deleted
 	if !watcherObj.DeletionTimestamp.IsZero() {
 		err := r.IstioClient.RemoveVirtualServiceConfigForCR(ctx, req.NamespacedName)
 		if err != nil {
-			return ctrl.Result{Requeue: true}, err
+			return ctrl.Result{}, err
 		}
 		controllerutil.RemoveFinalizer(watcherObj, watcherFinalizer)
 		err = r.Update(ctx, watcherObj)
 		if err != nil {
-			return ctrl.Result{Requeue: true}, err
+			return ctrl.Result{}, err
 		}
-		return ctrl.Result{}, nil
+		return ctrl.Result{Requeue: true}, nil
 	}
 
 	// check finalizer on native object
 	if !controllerutil.ContainsFinalizer(watcherObj, watcherFinalizer) {
 		controllerutil.AddFinalizer(watcherObj, watcherFinalizer)
 		err := r.Update(ctx, watcherObj)
 		if err != nil {
-			return ctrl.Result{Requeue: true}, err
+			return ctrl.Result{}, err
 		}
-		return ctrl.Result{}, nil
+		return ctrl.Result{Requeue: true}, nil
 	}
-	err := r.IstioClient.UpdateVirtualServiceConfig(ctx, watcherObj)
+	virtualService, err := r.IstioClient.GetVirtualService(ctx)
+	if apierrors.IsNotFound(err) {
+		if _, err := r.IstioClient.CreateVirtualService(ctx, watcherObj); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create virtual service %w", err)
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	err = r.IstioClient.UpdateVirtualServiceConfig(ctx, watcherObj, virtualService)
 	if err != nil {
-		return ctrl.Result{Requeue: true}, err
+		return ctrl.Result{}, err
 	}
```

Comment:
I think it's better to extract such domain-specific (and low-level) logic from the controller loop, focusing here only on error handling, requeue policy, finalizers, timestamps etc.
I would try to move this sequence into a method inside `istio/client`, as it used to be before.

Reply:
I completely agree with the idea that for the control loop, it should not contains low level details. But for this case, I don't think what you mentioned here should consider as low level, to me, the function this `istioClient` defined is already abstracted essential high-level policy, it's the policy this watcher control loop should follows (1. check if VirtualService exists, 2. update VirtualService config), low level detail has been already encapsulated. The problem here maybe this `istioClient` is not designed abstract enough (basically it can be a generic client, but not specific to istio), which leaves the impression that the dependency inversion principle is violated here.

Reply:
@ruanxin I still think this is too low level for the controller-loop.
IMO in the `Reconcile()` function we should only have something like: `ReconcileVirtualServices()` call (meaning: what to do), instead of embedding the necessary steps (meaning: how to do it), as it is now.

For example, the current implementation goes somewhat like this: `Get -> IfNotExistsCreate -> ElseUpdate`.
It could be replaced by, perhaps more efficient, ServerSideApply from K8s.
If we want to introduce such change, we would have to make a lot of changes in the Reconcile() implementation, which shouldn't be the case.
The knowledge of how we perform Istio configurations changes, and how many steps are necessary, shouldn't be a part of the controller loop logic.
At least that is my opinion on the correct responsibility split in the K8s operators, a general pattern.

Having said that, The code is readable and it's easy to refactor if necessary, so I can accept the PR.

Reply:
@ruanxin If we use SSA, then the "GET" step is not necessary. SSA works (kind of) like: create-or-update by design.